draft-davidben-tls-merkle-tree-certs.md

```diff
@@ -1100,6 +1110,23 @@ However, if the transparency service's interface becomes unavailable, monitors w
 
 {{trust-anchor-negotiation}} and {{tls-trust-anchors-extension}} involve the relying party sending a list of TrustAnchor values to aid the subscriber in selecting certificates. A sufficiently large list may be impractical to fit in a ClientHello and require alternate negotiation mechanisms or a different PKI structure. To reduce overhead, `issuer_id` values SHOULD be short, no more than eight bytes long.
 
+## Bitflips
+
+Bitflips happen at scale; indeed bitflips invalidated Certificate Transparency logs: see {{NESSIE2024BITFLIP}} and {{YETI2022BITFLIP}}.
+
+Where security allows it, the present system has been designed to be lenient with bitflips. For example:
+
+1. A signature that does not verify is not considered an misissuance, but rather an availability issue. This allows for bitflips in the signature itself and the signed message.
+2. An (abridged) assertion published via the HTTP interface, is ignored unless it's included in a valid signed window.
+3. A bitflip in a MerkleTreeProofSHA256 can be safely ignored, as any BikeshedCertificate that include it will fail to verify.
+
+There are several places where bitflips cannot be tolerated. For instance, misissuances include (but are not limited to)
+
+1. A bitflip in an assertion, which is included (with bitflip) in a valid signed window.
+2. A valid signed window, for which the CA cannot *eventually* produce assertions that lead to the included tree heads.
```

Comment:
Aren't the two cited bitflips really just analogs to these two failures? I.e., at least as currently written, we aren't any more bitflip-lenient than CT was. There is potentially some room for bitflip leniency purely on grounds that the bust trees will eventually age out, and you could imagine the TSes just refusing to forward the bust tree heads, but that feels like something different and may require some more normative text.

Reply:
> There is potentially some room for bitflip leniency purely on grounds that the bust trees will eventually age out, and you could imagine the TSes just refusing to forward the bust tree heads

It's scary to be lenient like this, because it threatens transparency. We don't want a CA to just show a particular batch to only one tiny TS.

> Aren't the two cited bitflips really just analogs to these two failures?

Yes and no.

The yeti2022 bit flip indeed is an example of 2, and we can't be lenient with it. The bit flip would've been caught, if the CT log would've recomputed the tree head just to be sure, as we suggest in the text.

The nessie2024 bit flip is a bit funny. It's not quite 1. The problem was that the SubjectPublicKeyInfo in the tbs_certificate didn't match the one in the pre_certificate because of a bit flip. With MTC there is no duplication: we'd just have a bit flip in the subject_info_hash. It's the subscriber's job to notice the error. If they see that hash only differs by one bit from the intended hash, then that's technically a misissuance, but one without any potential for a security breach, and we can be lenient.

Reply:
> It's scary to be lenient like this, because it threatens transparency. We don't want a CA to just show a particular batch to only one tiny TS.

Even the tiny TS will detect this because the input assertions will not match the signature at the top. Your TS, large or small, should be unwilling to pass tree heads along if they weren't able to build the tree.

But, yeah, it would be a whole lot of engineering and policy work when the alternative is pretty straightforward:
* The update process will naturally fail to work and so the CA -> TS -> RP flow will freeze. No new certs will work but old ones are still fine
* Recovery can simply be winding down the old CA and making a new one

> It's the subscriber's job to notice the error. If they see that hash only differs by one bit from the intended hash, then that's technically a misissuance, but one without any potential for a security breach, and we can be lenient.

I dunno how much we want to play the bit-flip-counting game here (also the bitflip could have happened in the middle of computing the hash, in which case it'll be a bit more of a mess). But my actual point is that nothing in this PR is special to MTCs.

* In both systems, if you thought you signed something but didn't, you didn't do anything. Although, more on this below.
* In both systems, if you serve some garbage instead of cryptographically coherent log state, it's an availability issue in your serving frontend
* In both systems, if you issue and log a certificate just fine, but send some other byte string to the subscriber, that's between you and the subscriber. Try again at delivering your byte string.
* In both systems, if you issue a certificate asserting a false thing, that's unambiguously misissuance. It's up to root programs to make a policy decision what to do here.
* In both systems, if you sign the wrong or a garbage tree head, that's unambiguously a failure in the log. MTCs can contain the impacts a bit better just because each batch is independent, but from the above discussion it sounds like we're not too excited about leaning on that anyway
* In both systems, double-checking computation can help if your computing environment is prone to bit flips

So I think we aren't _actually_ "designed to be lenient with bitflips". Which I also think is fine. It is hard to quantify the impact of a bitflip. Leaving it to a case-by-case policy decision seems the right path anyway. For example, the first claim ("A signature that does not verify...") is very very dangerous. If your bitflip caused you branch into the DoNothing() function instead of the SignSomething() function, that's fine. But more likely your bitflip messed up some internal computation in your signing algorithm, in which case we'd need to dig into the literature on fault attacks for your algorithm. You may well have burned your private key.

So I'm inclined to say we just don't say anything in the draft and leave it be.

Reply:
> Even the tiny TS will detect this because the input assertions will not match the signature at the top. Your TS, large or small, should be unwilling to pass tree heads along if they weren't able to build the tree.

💯 

> But, yeah, it would be a whole lot of engineering (punching holes in the update flow) and policy work when the alternative is pretty straightforward:
> 
> * The update process will naturally fail to work and so the CA -> TS -> RP flow will freeze. No new certs will work but old ones are still fine

This is sensible.

> * In both systems, double-checking computation can help if your computing environment is prone to bit flips

The nessie2024 bit flip would've been prevented if the computation was double checked. Whether some in an RFC will ensure all implementations will be thoughtful of bitflips is a different question.

> So I think we aren't _actually_ "designed to be lenient with bitflips".

I'm ok leaving that out. It's more important that we actually thought through where bitflips can happen and how we should deal with them.